android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt

```diff
@@ -0,0 +1,157 @@
+package com.badoo.ribs.android
+
+import android.annotation.TargetApi
+import android.content.pm.PackageManager
+import android.os.Build
+import android.support.v4.app.ActivityCompat
+import android.support.v4.content.ContextCompat
+import android.support.v7.app.AppCompatActivity
+import com.badoo.ribs.android.PermissionRequester.CheckPermissionsResult
+import com.badoo.ribs.android.PermissionRequester.RequestPermissionsEvent
+import com.badoo.ribs.android.PermissionRequester.RequestPermissionsEvent.Cancelled
+import com.badoo.ribs.android.PermissionRequester.RequestPermissionsEvent.RequestPermissionsResult
+import com.badoo.ribs.core.Identifiable
+import com.badoo.ribs.core.requestcode.RequestCodeRegistry
+import com.jakewharton.rxrelay2.PublishRelay
+import io.reactivex.Observable
+
+class PermissionRequesterImpl(
+    private val activity: AppCompatActivity,
+    private val requestCodeRegistry: RequestCodeRegistry
+) : PermissionRequester, PermissionRequestResultHandler {
+
+    private val events = HashMap<Int, HashMap<Int, PublishRelay<RequestPermissionsEvent>>>()
+
+    override fun checkPermissions(
+        client: Identifiable,
+        permissions: Array<String>
+    ) : CheckPermissionsResult {
+        val granted = mutableListOf<String>()
```

Comment:
I don't see a problem with using a mutable structure as an intermediate step. Also this makes three passes on the input instead of just one. Not that it's a big issue, but still. I'm open to hearing others though.

Reply:
I think mutable list here is more efficient and readable.
However, consider simplifying the logic inside, right now it is a bit hard to follow, extracting *Compat calls to extension functions for example.

Reply:
I like @Nublo approach more because for me it's more readable

Reply:
One more thought: using when looks a bit better:
```
permissions.forEach { permission ->
    when {
        permission.granted() -> granted += permission
        permission.showRationale() -> shouldShowRationale += permission
        else -> canAsk += permission
    }
}
```

Reply:
This approach is the easiest to understand so far, I'll go with it, thanks.

Reply:
Done.

Reply:
I like @Nublo suggestion, I find it more readable.
Three passes over the list of requested permission don't affect anyone, we can sacrifice that and increase readability.